src/main/java/hudson/plugins/mercurial/MercurialSCM.java

```diff
@@ -654,110 +642,99 @@ private void determineChanges(Run<?, ?> build, Launcher launcher, TaskListener l
+                    args.add("log");
+                    args.add("--template", MercurialChangeSet.CHANGELOG_TEMPLATE);
+                    if (revisionType == RevisionType.REVSET) {
+                        args.add("--rev", "ancestors(" + revToBuild + ") and not ancestors(" + prevTag.getId() + ")");
+                    } else {
+                        args.add("--rev", "ancestors('" + revToBuild.replace("'", "\\'") + "') and not ancestors(" + prevTag.getId() + ")");
+                    }
+                    args.add("--encoding", "UTF-8");
+                    args.add("--encodingmode", "replace");
 
-                ByteArrayOutputStream errorLog = new ByteArrayOutputStream();
+                    ByteArrayOutputStream errorLog = new ByteArrayOutputStream();
 
-                int r = hg.launch(args).stdout(new ForkOutputStream(os, errorLog)).pwd(repository).join();
-                if(r!=0) {
-                    Util.copyStream(new ByteArrayInputStream(errorLog.toByteArray()), listener.getLogger());
-                    throw new IOException("Failure detected while running hg log to determine change log");
+                    int r = hg.launch(args).stdout(new ForkOutputStream(os, errorLog)).pwd(repository).join();
+                    if (r != 0) {
+                        Util.copyStream(new ByteArrayInputStream(errorLog.toByteArray()), listener.getLogger());
+                        throw new IOException("Failure detected while running hg log to determine change log");
+                    }
+                } finally {
+                    os.write("</changesets>".getBytes("UTF-8"));
                 }
-            } finally {
-                os.write("</changesets>".getBytes("UTF-8"));
             }
-        } finally {
-            os.close();
-        }
-        } finally {
-            hg.close();
         }
     }
 
     private void update(Run<?, ?> build, Launcher launcher, FilePath repository, Node node, TaskListener listener, String toRevision, StandardUsernameCredentials credentials)
             throws IOException, InterruptedException {
-        HgExe hg = new HgExe(findInstallation(getInstallation()), credentials, launcher, node, listener, build.getEnvironment(listener));
         EnvVars env = build.getEnvironment(listener);
-        try {
-        int pullExitCode;
-        try {
-            pullExitCode = pull(launcher, repository, listener, node, toRevision, credentials, env);
-        } catch (IOException e) {
-            if (causedByMissingHg(e)) {
-                listener.error("Failed to pull because hg could not be found;" +
-                        " check that you've properly configured your Mercurial installation");
-            } else {
-                e.printStackTrace(listener.error("Failed to pull"));
+        try (HgExe hg = new HgExe(findInstallation(getInstallation()), credentials, launcher, node, listener, build.getEnvironment(listener))) {
```

Comment:
```suggestion
        try (HgExe hg = new HgExe(findInstallation(getInstallation()), credentials, launcher, node, listener, env)) {
```

Reply:
Good catch. I've found another occasion. I just updated my commit accordingly.

Reply:
BTW when you force-push it prevents me from using GH’s feature of reviewing only new commits since the last review; I have to start over from scratch.

Reply:
Sorry for the extra effort. My goal most of the time is to keep the history as clean as possible.

Reply:
My philosophy FWIW:

* If a repo maintainer wishes to have a linear commit history in the origin repo, they are always free to squash-merge. There is no need to destroy history inside the PR branch.
* Even if plain merges are used and the repo contains a complex history inside those branches, it is trivial to suppress those details with `--first-parent` or the like; but you cannot undo the loss of information.